presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java

```diff
@@ -0,0 +1,393 @@
+
+        protected DereferencePushDownRule(Metadata metadata, SqlParser sqlParser, Pattern<N> targetPattern)
+        {
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+            this.targetPattern = requireNonNull(targetPattern, "targetPattern is null");
+        }
+
+        @Override
+        public Pattern<ProjectNode> getPattern()
+        {
+            return project().with(source().matching(targetPattern.capturedAs(targetCapture)));
+        }
+
+        @Override
+        public Result apply(ProjectNode node, Captures captures, Context context)
+        {
+            N child = captures.get(targetCapture);
+            Map<DereferenceExpression, Symbol> expressions = getDereferenceSymbolMap(node, context, metadata, sqlParser);
+            Assignments assignments = node.getAssignments().rewrite(new DereferenceReplacer(expressions));
+
+            Result result = pushDownDereferences(context, child, expressions, assignments);
+            if (result.isEmpty()) {
+                return Result.empty();
+            }
+            return Result.ofPlanNode(new ProjectNode(context.getIdAllocator().getNextId(), result.getTransformedPlan().get(), assignments));
+        }
+
+        protected abstract Result pushDownDereferences(Context context, N targetNode, Map<DereferenceExpression, Symbol> expressions, Assignments assignments);
+
+        protected boolean isCastRowType(DereferenceExpression dereference)
+        {
+            Expression base = dereference.getBase();
+            if (base instanceof Cast) {
+                return metadata.getTypeManager().getType(parseTypeSignature(((Cast) base).getType())) instanceof RowType;
+            }
+            return false;
+        }
+    }
+
+/**
+ * Transforms:
+ * <pre>
+ *  Project(a.msg.x)
+ *    Join(a.msg.y = b.msg.y)
+ *      Project(a.msg.y)
+ *          Source(a)
+ *      Project(b.msg.y)
+ *          Source(b)
+ *  </pre>
+ * to:
+ * <pre>
+ *  Join(a.msg.y = b.msg.y)
+ *    Project(a.msg.x, a.msg.y)
+ *      Source(a)
+ *    Project(b.msg.y)
+ *      Source(b)
+ *  </pre>
+ */
+    public class DereferencePushDownJoin
+            extends DereferencePushDownRule<JoinNode>
+    {
+        public DereferencePushDownJoin(Metadata metadata, SqlParser sqlParser)
+        {
+            super(metadata, sqlParser, join());
+        }
+
+        @Override
+        protected Result pushDownDereferences(Context context, JoinNode joinNode, Map<DereferenceExpression, Symbol> expressions, Assignments assignments)
+        {
+            List<Symbol> outputSymbols = joinNode.getOutputSymbols();
+            Map<Symbol, Expression> pushdownExpressions = expressions.entrySet().stream()
+                    .filter(entry -> !isCastRowType(entry.getKey()))
+                    .filter(entry -> outputSymbols.contains(getOnlyElement(extractAll(entry.getKey()))))
```

Comment:
Is is not possible for a dereference expression to use more then one symbol, e.g. `f(a, b).x`?